File: minigit-4/py/commands/update_ref.py
"""update-ref command - Update a reference."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from refs import set_ref


def run(args):
    """Update a reference."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    if len(args) < 2:
        print("usage: minigit update-ref <ref> <sha>", file=sys.stderr)
        return 1

    ref_name = args[0]
    sha = args[1]

    # Validate SHA format
    if len(sha) < 4 or not all(c in '0123456789abcdef' for c in sha):
        print(f"fatal: invalid sha: {sha}", file=sys.stderr)
        return 1

    set_ref(ref_name, sha, repo_root)
    return 0

File: minigit-4/ts/src/commands/update-ref.ts
// update-ref command - Update reference

import { findRepoRoot } from '../utils';
import { updateRef, writeHead } from '../refs';

export function updateRefCmd(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  if (args.length < 2) {
    console.error('fatal: update-ref requires <ref> <sha>');
    return 1;
  }

  const refName = args[0];
  const sha = args[1];

  // Validate SHA (should be 40 hex chars or abbreviated)
  if (!/^[0-9a-f]{4,40}$/.test(sha)) {
    console.error(`fatal: ${sha} is not a valid SHA`);
    return 1;
  }

  if (refName === 'HEAD') {
    writeHead(repoRoot, sha);
  } else {
    updateRef(repoRoot, refName, sha);
  }

  return 0;
}
